### PR TITLE
Competitive Wind Farm Exclusions

### PR DESCRIPTION
**Pull request by MRossol**
Implementation of #59 

Add CompetitiveWindFarms class to supply_curve
CompetitiveWindFarms will handle the upwind/downwind wind farm exclusion and can exclude upwind/downwind neighbors from a sc_point table

Integrated CompetitiveWindFarms into SupplyCurve full and simple sort

added tests

### CODE REVIEW

File: tests/test_supply_curve_wind_dirs.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+    if downwind:
+        baseline = os.path.join(TESTDATADIR, 'comp_wind_farms',
+                                'sc_full_downwind.csv')
+    else:
+        baseline = os.path.join(TESTDATADIR, 'comp_wind_farms',
+                                'sc_full_upwind.csv')
+
+    baseline = pd.read_csv(baseline)
+    assert_frame_equal(sc_out, baseline, check_dtype=False)
```
Comment (grantbuster) [inline tests/test_supply_curve_wind_dirs.py:62]: Did you verify the results? This gets back to the verification vs. baseline validation discussion. Always a bit nervous to see no code inspection of the results.

File: reV/supply_curve/supply_curve.py [hunk 2]
```diff
@@ -554,43 +862,80 @@ def _calculate_total_lcoe_friction(self):
...
+        downwind : bool, optional
+            Flag to remove downwind neighbors as well as upwind neighbors,
+            by default False
+
+        Returns
+        -------
+        wind_dirs : CompetitiveWindFarms
+            updated CompetitiveWindFarms instance
+        """
+        gid = wind_dirs['sc_gid', sc_gid]
```
Comment (grantbuster) [inline reV/supply_curve/supply_curve.py:884]: I had a similar comment in the getter: what operation is this performing? Just the the variable names I have absolutely no idea.

File: reV/supply_curve/supply_curve.py [hunk 3]
```diff
@@ -19,6 +19,314 @@
 logger = logging.getLogger(__name__)
 
 
+class CompetitiveWindFarms:
```
Comment (grantbuster) [inline reV/supply_curve/supply_curve.py:22]: Move this into its own file? This one is breaking 1200 lines.

File: reV/supply_curve/supply_curve.py [hunk 4]
```diff
@@ -554,43 +862,80 @@ def _calculate_total_lcoe_friction(self):
...
+            upwind_gids = wind_dirs['upwind', gid]
+            for n in upwind_gids:
+                check = wind_dirs.mask_gid(n)
+                if check:
+                    sc_gids = wind_dirs['sc_point_gid', n]
+                    for sc_id in sc_gids:
+                        self._mask[sc_id] = False
+
+        if downwind:
+            downwind_gids = wind_dirs['downwind', gid]
```
Comment (grantbuster) [inline reV/supply_curve/supply_curve.py:895]: You use all four of the `wind_dirs.__getitem__` options here but its not clear what is being input/output. There's so much confusion between sc_gid and sc_point_gid, and of course all the other gids (which I know is not the case here, but still). I would suggest splitting the getitem into four explicit methods.

**Comment by MRossol** [comment]
I haven't run a full sc test yet as I'm fighting to get the wind_dirs file made. process pool workers keep dying, I think due to lack of memory... I've reduce the max_workers to 8...

File: reV/supply_curve/supply_curve.py [hunk 5]
```diff
@@ -19,6 +19,314 @@
...
+                                                           n_dirs=n_dirs)
+
+    def __repr__(self):
+        shape = self._upwind.shape
+        msg = ("{} with {} sc_point_gids and {} prominent directions"
+               .format(self.__class__.__name__, shape[0], shape[1]))
+
+        return msg
+
+    def __getitem__(self, keys):
```
[the remark this answers is not in the document]
Comment (MRossol) [reply reV/supply_curve/supply_curve.py:56]: fair and done, added methods for each and docstrings to the getitem

File: reV/supply_curve/supply_curve.py [hunk 6]
```diff
@@ -19,6 +19,314 @@
...
+        if not isinstance(keys, tuple):
+            msg = ("{} must be a tuple of form (source, gid) where source is: "
+                   "'sc_gid', 'sc_point_gid',  or 'upwind', 'downwind'"
+                   .format(keys))
+            logger.error(msg)
+            raise ValueError(msg)
+
+        source, gid = keys
+        if source == 'sc_gid':
+            out = self._sc_gids.loc[gid].values[0]
```
[the remark this answers is not in the document]
Comment (MRossol) [reply reV/supply_curve/supply_curve.py:66]: see new sub-methods

File: reV/supply_curve/supply_curve.py [hunk 7]
```diff
@@ -19,6 +19,314 @@
...
+                   "'downwind'".format(source))
+            logger.error(msg)
+            raise ValueError(msg)
+
+        return out
+
+    @property
+    def mask(self):
+        """
+        Supply curve point boolean mask, used for efficient exclusion
```
[the remark this answers is not in the document]
Comment (MRossol) [reply reV/supply_curve/supply_curve.py:84]: Done

File: reV/supply_curve/supply_curve.py [hunk 8]
```diff
@@ -19,6 +19,314 @@
...
+        downwind_gids = np.take_along_axis(downwind_gids, neighbors, axis=1)
+        downwind_gids = pd.DataFrame(downwind_gids,
+                                     index=wind_dirs.index.values,
+                                     columns=[i for i in range(n_dirs)])
+
+        return upwind_gids, downwind_gids
+
+    def mask_gid(self, sc_point_gid):
+        """
+        Mask given supply curve point id
```
[the remark this answers is not in the document]
Comment (MRossol) [reply reV/supply_curve/supply_curve.py:222]: ugg, I struggled with that one... changed to exclude_sc_point_gid

File: reV/supply_curve/supply_curve.py [hunk 9]
```diff
@@ -554,43 +862,80 @@ def _calculate_total_lcoe_friction(self):
...
-                           'trans_cap_cost', 'dist_mi', 'lcot', 'total_lcoe')):
+    def _competitive_wind_farms(self, wind_dirs, sc_gid, downwind=False):
         """
-        run supply curve sorting in serial
+        Exclude non-competitive wind farms for given sc_gid
 
         Parameters
         ----------
-        trans_table : pandas.DataFrame | NoneType
+        wind_dirs : CompetitiveWindFarms
```
[the remark this answers is not in the document]
Comment (MRossol) [reply reV/supply_curve/supply_curve.py:871]: Changed to comp_wind_dirs

File: reV/supply_curve/supply_curve.py [hunk 10]
```diff
@@ -554,43 +862,80 @@ def _calculate_total_lcoe_friction(self):
             logger.info('Found mean LCOE with friction. Adding key '
                         '"total_lcoe_friction" to trans table.')
 
-    def full_sort(self, trans_table=None, sort_on='total_lcoe',
-                  columns=('trans_gid', 'trans_capacity', 'trans_type',
-                           'trans_cap_cost', 'dist_mi', 'lcot', 'total_lcoe')):
+    def _competitive_wind_farms(self, wind_dirs, sc_gid, downwind=False):
```
[the remark this answers is not in the document]
Comment (MRossol) [reply reV/supply_curve/supply_curve.py:865]: damn naming... done

File: reV/supply_curve/supply_curve.py (hunk 2 above)
Comment (MRossol) [reply reV/supply_curve/supply_curve.py:884]: see new method

File: reV/supply_curve/supply_curve.py [hunk 11]
```diff
@@ -634,26 +979,96 @@ def full_sort(self, trans_table=None, sort_on='total_lcoe',
...
+        pos = trans_table['lcot'].isnull()
+        trans_table = trans_table.loc[~pos].sort_values(sort_on)
+
+        total_lcoe_fric = None
+        if self._consider_friction and 'mean_lcoe_friction' in trans_table:
+            columns.append('total_lcoe_friction')
+            total_lcoe_fric = trans_table['total_lcoe_friction'].values
+
+        if wind_dirs is not None:
+            wind_dirs = CompetitiveWindFarms(wind_dirs, self._sc_points,
```
[the remark this answers is not in the document]
Comment (MRossol) [reply reV/supply_curve/supply_curve.py:1059]: What do you mean? the arg name or the internal variable?

File: tests/test_supply_curve_wind_dirs.py (hunk 1 above)
Comment (MRossol) [reply tests/test_supply_curve_wind_dirs.py:62]: I don't disagree, but any thoughts on better tests?

File: reV/supply_curve/supply_curve.py (hunk 3 above)
Comment (MRossol) [reply reV/supply_curve/supply_curve.py:22]: done

File: reV/supply_curve/supply_curve.py (hunk 4 above)
Comment (MRossol) [reply reV/supply_curve/supply_curve.py:895]: see above(s)

File: reV/supply_curve/supply_curve.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (grantbuster) [reply reV/supply_curve/supply_curve.py:1059]: Both kind of. This makes sense: https://github.com/NREL/reV/blob/86e7797e60b74e8af7517bb40147e886ed2bf9f7/reV/supply_curve/supply_curve.py#L1028-L1031

This does not make sense: https://github.com/NREL/reV/blob/86e7797e60b74e8af7517bb40147e886ed2bf9f7/reV/supply_curve/supply_curve.py#L917-L918

When the wind_dirs internal variable gets re-assigned at 1059 as a CompetitiveWindFarm instance, the variable name no longer becomes accurate. We should use a second variable that is the instance of the CompetitiveWindFarms object and leave wind_dirs un-changed.

File: tests/test_supply_curve_wind_dirs.py (hunk 1 above)
Comment (grantbuster) [reply tests/test_supply_curve_wind_dirs.py:62]: Could we go through the built SC table and check the upwind/downwind gid's and make sure they are not present in the built SC table? Just confirm that they were indeed competitively eliminated.

File: reV/supply_curve/supply_curve.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (MRossol) [reply reV/supply_curve/supply_curve.py:1059]: Done, but I think this is revenge for all of the line breaks ;-)

File: tests/test_supply_curve_wind_dirs.py (hunk 1 above)
Comment (MRossol) [reply tests/test_supply_curve_wind_dirs.py:62]: I like it, will do

**Comment by MRossol** [comment]
@grantbuster  FYI, I changed the getitem meanings, now it should be clearer self[item to get, source gid],  so comp_wind_dirs['sc_point_gid', sc_gid] will get the sc_point_gid that maps to the given sc_gid, much clearer to the user I hope

File: reV/supply_curve/supply_curve.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (grantbuster) [reply reV/supply_curve/supply_curve.py]: haha no way you gotta keep those variable names accurate and descriptive! :)

**Comment by grantbuster** [comment]
@MRossol, looks good, thanks for responding to comments. It looks like you've added that suggested test and maybe are trying to get it to run. I'll approve once its sorted. I wouldn't be surprised if we have performance issues with the added complexity, but the technical stuff looks good.

File: reV/supply_curve/supply_curve.py [hunk 12]
```diff
@@ -554,43 +555,81 @@ def _calculate_total_lcoe_friction(self):
...
+
+        return comp_wind_dirs
+
+    def _full_sort(self, trans_table, wind_dirs, total_lcoe_fric,
+                   sort_on='total_lcoe',
+                   columns=('trans_gid', 'trans_capacity', 'trans_type',
+                            'trans_cap_cost', 'dist_mi', 'lcot', 'total_lcoe'),
+                   downwind=False):
+        """
+        [summary]
```
Comment (grantbuster) [inline reV/supply_curve/supply_curve.py:605]: update doc strings

File: reV/supply_curve/supply_curve.py [hunk 13]
```diff
@@ -554,43 +555,81 @@ def _calculate_total_lcoe_friction(self):
...
+
+        Parameters
+        ----------
+        trans_table : pandas.DataFrame
             Supply Curve Tranmission table to sort on
-            If none use self._trans_table
-        sort_on : str
+        wind_dirs : CompetitiveWindFarms
```
Comment (grantbuster) [inline reV/supply_curve/supply_curve.py]: Shouldn't this be an optional kwarg?

File: reV/supply_curve/supply_curve.py [hunk 14]
```diff
@@ -634,26 +673,97 @@ def full_sort(self, trans_table=None, sort_on='total_lcoe',
...
+        pos = trans_table['lcot'].isnull()
+        trans_table = trans_table.loc[~pos].sort_values(sort_on)
+
+        total_lcoe_fric = None
+        if self._consider_friction and 'mean_lcoe_friction' in trans_table:
+            columns.append('total_lcoe_friction')
+            total_lcoe_fric = trans_table['total_lcoe_friction'].values
+
+        if wind_dirs is not None:
+            comp_wind_dirs = CompetitiveWindFarms(wind_dirs, self._sc_points,
```
Comment (grantbuster) [inline reV/supply_curve/supply_curve.py:753]: I think comp_wind_dirs should be set to None above the if statement and then have it be a kwarg for the protected method _full_sort

File: reV/supply_curve/supply_curve.py (hunk 12 above)
Comment (MRossol) [reply reV/supply_curve/supply_curve.py]: done

File: reV/supply_curve/supply_curve.py (hunk 13 above)
Comment (MRossol) [reply reV/supply_curve/supply_curve.py]: agreed and fixed

File: reV/supply_curve/supply_curve.py (hunk 14 above)
Comment (MRossol) [reply reV/supply_curve/supply_curve.py:754]: yup, see above

**Pull request merged**